ssl/record/rec_layer_s3.c

```diff
@@ -296,6 +296,12 @@ int ssl3_read_n(SSL *s, size_t n, size_t max, int extend, int clearold,
             ret = BIO_read(s->rbio, pkt + len + left, max - left);
             if (ret >= 0)
                 bioread = ret;
+            if (ret <= 0
+                    && !BIO_should_retry(s->rbio)
+                    && BIO_eof(s->rbio)) {
```

Comment:
Why isn't the "should retry" test enough?

Reply:
Because if we haven't hit eof  then we want to be returning SSL_ERROR_SYSCALL and have the application check errno to discover the problem. However if we hit eof then errno is not set do we want to add our own error to the stack (in order to return SSL_ERROR_SSL)